torch_geometric/sampler/neighbor_sampler.py

```diff
@@ -396,106 +389,237 @@ def edge_sample(
-            seed_nodes, inverse = seed_nodes.unique(return_inverse=True)
-            edge_label_index = inverse.view(2, -1)
+        seed = torch.cat([src, dst], dim=0)
+        seed_time = None
 
-        output = sample_fn(seed=seed_nodes, seed_time=seed_time)
+        if not disjoint:
+            seed, inverse_seed = seed.unique(return_inverse=True)
 
-        if disjoint:
-            output.batch = output.batch % num_seed_edges
+        if edge_label_time is not None:  # Always disjoint.
+            seed_time = torch.cat([src_time, dst_time])
 
-        output.metadata = (index, edge_label_index, edge_label,
-                           edge_label_time)
+        out = sample_fn(seed=seed, seed_time=seed_time)
 
-    return output
+        # Enhance `out` by label information ##################################
+        if neg_sampling is None or neg_sampling.is_binary():
+            if disjoint:
+                out.batch = out.batch % num_pos
+                edge_label_index = torch.arange(2 * seed.numel()).view(2, -1)
+            else:
+                edge_label_index = inverse_seed.view(2, -1)
+
+            out.metadata = (index, edge_label_index, edge_label, seed_time)
+
+        elif neg_sampling.is_triplet():
+            if disjoint:
+                out.batch = out.batch % num_pos
+                src_index = torch.arange(num_pos)
+                dst_pos_index = torch.arange(num_pos, 2 * num_pos)
+                # `dst_neg_index` needs to be offset such that indices with
+                # offset `num_pos` belong to the same triplet:
+                dst_neg_index = torch.arange(2 * num_pos, seed.numel())
+                dst_neg_index = dst_neg_index.view(-1, num_pos).t()
+            else:
+                src_index = inverse_seed[:num_pos]
+                dst_pos_index = inverse_seed[num_pos:2 * num_pos]
+                dst_neg_index = inverse_seed[2 * num_pos:]
+            dst_neg_index = dst_neg_index.view(num_pos, -1).squeeze(-1)
+
+            out.metadata = (index, src_index, dst_pos_index, dst_neg_index,
+                            src_time)
+
+    return out
+
+
+def neg_sample(seed: Tensor, num_samples: int, num_nodes: int,
+               seed_time: Optional[Tensor],
+               node_time: Optional[Tensor]) -> Tensor:
+    if node_time is None:
+        return torch.randint(num_nodes, (seed.numel() * num_samples, ))
+
+    # Otherwise, we can only sample negatives that exist in the graph:
+    # TODO See if this greedy algorithm here can be improved.
+    assert seed_time is not None
+    seed_time = seed_time.view(1, -1).expand(num_samples, -1)
+    out = torch.randint(num_nodes, (num_samples, seed.numel()))
+    mask = node_time[out] >= seed_time
+    neg_sampling_complete = False
+    for i in range(5):  # pragma: no cover
+        if not mask.any():
+            neg_sampling_complete = True
+            break
+
+        # Greedily search for alternative negatives.
```

Comment:
is there some way to avoid sampling false negatives, given we have `edge_label_index`

Reply:
Will add a TODO for this. I think this is currently out-of-scope for this PR.